app/services/puppet_fact_parser.rb

```diff
@@ -6,8 +6,8 @@ def operatingsystem
 
     if orel.present?
       major, minor = orel.split('.', 2)
-      major = major.to_s.gsub(/\D/, '')
-      minor = minor.to_s.gsub(/[^\d\.]/, '')
+      major = major.to_s.gsub(/\D/, '') || ''
+      minor = minor.to_s.gsub(/[^\d.]/, '') || ''
```

Comment:
I don't think changing `\.` to `.` is correct. If it is, you change it to `/[^.]/` or essentially `.` so it will always drop the character.

Funny thing: the git blame for this line is 13875289357605c80b8073c2befcf84303e3695d and it speaks about `minor` being `nil` instead of `""`. I vaguely recall that `nil.to_s` returns an empty string instead of `nil`. We probably rely on this Ruby behavior to ensure non-null entries.

```
[1] pry(main)> nil.to_s
=> ""
```

That means this code is not the offender, right?

The problem does not show up with vanilla Foreman, only with Katello. Perhaps this is coming from the RHSM fact parser? However, that has this code:
https://github.com/theforeman/foreman/blob/c693cd93ad80e2374dd42f50aaeff5225e6b9f91/app/services/katello/rhsm_fact_parser.rb#L60

I would think the `|| ''` would take care of it.

So I'm not entirely sure where this comes from. I'll see if I can spin up a pipeline for a reproducer.

Reply:
As I said in https://community.theforeman.org/t/katello-nightly-rpm-pipeline-1220-failed/27082/11: currently nightly is passed and I can no longer find a reproducer. That makes it very hard to review this. However, since `nil.to_s` is already `""`, the `|| ''` part will never be triggered and I think it can be left out.

Reply:
> I don't think changing \. to . is correct 

This was recommended by me by Rubocop: https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/RedundantRegexpEscape

What is your concern? I don't understand.

> currently nightly is passed and I can no longer find a reproducer

Well since we are at it and the code with `to_s` really smells, let's try to ensure there are no nil values via the migration and constraint rule? In the worst case, well, we break this. Not for the first time or last...

Reply:
> This was recommended by me by Rubocop: https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/RedundantRegexpEscape

Interesting. I didn't think it was redundant.

> What is your concern? I don't understand.

I'm worried it'll be the catchall `.` instead of a literal `.`. However, testing suggests it is indeed the literal `.`:
```
[1] pry(main)> '2.1a'.to_s.gsub(/[^\d\.]/, '')
=> "2.1"
[2] pry(main)> '2.1a'.to_s.gsub(/[^\d.]/, '')
=> "2.1"
```

I must say I really dislike regexes and all these subtle nuances.